src/tactic/linarith.lean

```diff
@@ -98,6 +98,103 @@ by rw [←h3, mul_assoc, mul_div_comm, h2, ←mul_assoc, h1, mul_comm, one_mul]
 
 end lemmas
 
+/--
+A linear expression is a list of pairs of variable indices and coefficients.
+
+Some functions on `linexp` assume that `n : ℕ` occurs at most once as the first element of a pair,
+and that the list is sorted in decreasing order of the first argument.
+This is not enforced by the type but the operations here preserve it.
+-/
+@[reducible]
+def linexp : Type := list (ℕ × ℤ)
+end linarith
+
+/--
+A map `ℕ → ℤ` is converted to `list (ℕ × ℤ)` in the obvious way.
+This list is sorted in decreasing order of the first argument.
+-/
+meta def native.rb_map.to_linexp (m : rb_map ℕ ℤ) : linarith.linexp :=
+m.to_list
+
+namespace linarith
+namespace linexp
+
+/--
+Add two `linexp`s together componentwise.
+Preserves sorting and uniqueness of the first argument.
+-/
+def add : linexp → linexp → linexp
+| [] a := a
+| a [] := a
+| (a@(n1,z1)::t1) (b@(n2,z2)::t2) :=
+  match cmp n1 n2 with
+  | ordering.lt := b::add (a::t1) t2
+  | ordering.gt := a::add t1 (b::t2)
+  | ordering.eq := let sum := z1 + z2 in if sum = 0 then add t1 t2 else (n1, sum)::add t1 t2
+  end
+
+/-- `l.scale c` scales the values in `l` by `c` without modifying the order or keys. -/
+def scale (c : ℤ) (l : linexp) : linexp :=
+if c = 0 then []
+else l.map $ λ ⟨n, z⟩, (n, z*c)
+
+/--
+`l.get n` returns the value in `l` associated with key `n`, if it exists, and `none` otherwise.
+This function assumes that `l` is sorted in decreasing order of the first argument,
+that is, it will return `none` as soon as it finds a key smaller than `n`.
+-/
+def get (n : ℕ) : linexp → option ℤ
+| [] := none
+| ((a, b)::t) := match cmp a n with
```

Comment:
BTW, feel free to use `if a < n then ...` if you think that improves readability.  It's also much faster than `cmp`.

Reply:
My intention using `cmp` was to avoid doing the comparison twice (i.e. checking a < b and b < a to conclude a = b). But I clearly should have looked at the definition of `cmp` first, since it doesn't do what I'd expect. Do we have a function/typeclass to efficiently produce `ordering`s?

Reply:
I guess this isn't so important for natural number comparisons.